### PR TITLE
Create custom pyspark image

### PR DESCRIPTION
**Pull request by Sondergaard**
## Description

Create a pyspark image to be used by wholesale.
This version is trimmed and is about half the size of the one used before.

## Quality

- [x] Documentation is updated
- `n/a` Release notes are updated
- `n/a` Package version is updated
- `n/a` Public types and methods are documented
- `n/a` Tests are implemented and executed locally

### CODE REVIEW

File: source/Images/pyspark-slim/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+ARG REGISTRY=quay.io
+ARG OWNER=jupyter
+ARG BASE_CONTAINER=$REGISTRY/$OWNER/base-notebook
+FROM $BASE_CONTAINER
+
+LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"
```
Comment (BjarkeMeier) [inline source/Images/pyspark-slim/Dockerfile:8]: Are we part of the Jupyter Project? 😉

File: source/Images/pyspark-slim/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# Fix: https://github.com/hadolint/hadolint/wiki/DL4006
+# Fix: https://github.com/koalaman/shellcheck/wiki/SC3014
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
```
Comment (BjarkeMeier) [inline source/Images/pyspark-slim/Dockerfile:12]: As we're not piping anything we could simplify by removing this.

File: source/Images/pyspark-slim/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+# 1. Check out the Spark branch you are on: <https://github.com/apache/spark>
+# 2. Find the pandas version in the file `dev/infra/Dockerfile`.
+RUN mamba install --yes \
+    'grpcio-status' \
+    'grpcio' \
+    'pandas=2.0.3' \
+    'pyarrow' && \
+    mamba clean --all -f -y && \
+    fix-permissions "${CONDA_DIR}" && \
+    fix-permissions "/home/${NB_USER}"
```
Comment (BjarkeMeier) [inline source/Images/pyspark-slim/Dockerfile:71]: @johevemi , do we need pandas or pyarrow? AFAIK I removed the last usage of pandas.

File: source/Images/pyspark-slim/setup_spark.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    LOGGER.info("Downloading Spark versions information")
+    all_refs = get_all_refs("https://archive.apache.org/dist/spark/")
+    stable_versions = [
+        ref.removeprefix("spark-").removesuffix("/")
+        for ref in all_refs
+        if ref.startswith("spark-") and "incubating" not in ref and "preview" not in ref
+    ]
+    # Compare versions semantically
+    latest_version = max(
+        stable_versions, key=lambda ver: [int(sub_ver) for sub_ver in ver.split(".")]
```
Comment (BjarkeMeier) [inline source/Images/pyspark-slim/setup_spark.py:56]: Congratulations on becoming a python developer. We have reserved a spot for you in team Mandalorian 😄

File: source/Images/pyspark-slim/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+# Setup Spark
+RUN /opt/setup-scripts/setup_spark.py \
+    --spark-version="${spark_version}" \
+    --hadoop-version="${hadoop_version}" \
+    --scala-version="${scala_version}" \
+    --spark-download-url="${spark_download_url}"
+
+# Configure IPython system-wide
+COPY ipython_kernel_config.py "/etc/ipython/"
+RUN fix-permissions "/etc/ipython/"
```
[the remark this answers is not in the document]
Comment (johevemi) [reply source/Images/pyspark-slim/Dockerfile:53]: Never heard of IPython, so i don't think we are using it

File: source/Images/pyspark-slim/Dockerfile (hunk 3 above)
Comment (johevemi) [reply source/Images/pyspark-slim/Dockerfile:71]: We are not using pandas or pyarrow anymore

File: source/Images/pyspark-slim/Dockerfile (hunk 2 above)
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:12]: Removed

File: source/Images/pyspark-slim/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+# Spark dependencies
+# Default values can be overridden at build time
+# (ARGS are in lowercase to distinguish them from ENV)
+ARG openjdk_version="17"
+
+RUN apt-get update --yes && \
+    apt-get install --yes --no-install-recommends \
+    "openjdk-${openjdk_version}-jre-headless" \
+    curl \
```
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:18]: curl is used within setup_spark.py and must be installed.

File: source/Images/pyspark-slim/setup_spark.py (hunk 4 above)
Comment (Sondergaard) [reply source/Images/pyspark-slim/setup_spark.py:56]: I won't take credit for copy paste work :)

File: source/Images/pyspark-slim/Dockerfile (hunk 3 above)
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:65]: Pandas and pyarrow removed from image

File: source/Images/pyspark-slim/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:47]: It seems like it is part of the [Jupyter part](https://ipython.org). I will leave the configuration as part of the docker image.

[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/Images/pyspark-slim/Dockerfile:47]: I don't think we use Jupyter either.

File: source/Images/pyspark-slim/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:18]: For now the focus is on getting the base image to work - we can then optimize.

File: source/Images/pyspark-slim/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sondergaard) [reply source/Images/pyspark-slim/Dockerfile:47]: It comes as a part of the base image. Main focus is to get the image working, then further configurations if needed.

**Comment by Sondergaard** [comment]
Closing PR - Branch is moved to python package repository.

**Pull request closed without merging**